hw/misc/pcie-testdev.c

```diff
@@ -0,0 +1,202 @@
+    DEFINE_PROP_UINT64("size", PCIeTestDevState, regs.size, false),
+    DEFINE_PROP_UINT64("busy_delay_ms", PCIeTestDevState, regs.busy_delay, false),
+    DEFINE_PROP_END_OF_LIST()
+};
+
+static uint64_t
+pcie_testdev_read(void *opaque, hwaddr addr, unsigned size)
+{
+    PCIeTestDevState *s = opaque;
+
+    switch(addr) {
+    case SRCADDR:
+        return s->regs.srcaddr;
+        break;
+    case DSTADDR:
+        return s->regs.dstaddr;
+        break;
+    case SIZE:
+        return s->regs.size;
+        break;
+    case STATUS:
+        return s->regs.status;
+        break;
+    case BUSY_DELAY:
+        return s->regs.busy_delay;
+    default:
+        return 0x0;
+    }
+}
+
+static void
+enable_bus_master(PCIeTestDevState *s){
+    pci_set_word(s->parent_obj.config + PCI_COMMAND,
+                 s->parent_obj.config[PCI_COMMAND] | PCI_COMMAND_MASTER);
+}
+
+static void
+dma_transfer(PCIeTestDevState *s){
+    s->regs.status = 0x01;
+    printf("DMA Transfer from source addr %p to dest addr %p, length %lu\n", (void *)s->regs.srcaddr, (void *)s->regs.dstaddr, s->regs.size);
+    uint16_t *buf = g_malloc(s->regs.size);
+    printf("Read from source, STATUS: %d\n", pci_dma_read(&s->parent_obj, s->regs.srcaddr, buf, s->regs.size));
+    printf("Write to dest, STATUS: %d\n", pci_dma_write(&s->parent_obj, s->regs.dstaddr, buf, s->regs.size));
+    g_free(buf);
+    s->regs.status = 0x00;
+
+    if (s->aggressive) {
+        QEMUTimer *dma_timer = timer_new_ms(QEMU_CLOCK_VIRTUAL, (void *) dma_transfer, s);
+        timer_mod(dma_timer, qemu_clock_get_ms(QEMU_CLOCK_VIRTUAL) + s->aggression_delay);
+    }
+}
+
+static void
+pcie_testdev_write(void *opaque, hwaddr addr, uint64_t val,
+                   unsigned size)
+{
+    PCIeTestDevState *s = opaque;
+
+    switch(addr) {
+    case SRCADDR:
+        s->regs.srcaddr = val;
+        break;
+    case DSTADDR:
+        s->regs.dstaddr = val;
+        break;
+    case SIZE:
+        s->regs.size = val;
+        break;
+    case DOORBELL:
+        s->regs.doorbell = val;
+        if (val & 0x01 && s->regs.size) {
+            s->regs.status = 0x01;
+
+            /* Use a timer to simulate delay in device making a DMA_transfer */
+            QEMUTimer *timer = timer_new_ms(QEMU_CLOCK_VIRTUAL, (void *) dma_transfer, s);
+            timer_mod(timer, qemu_clock_get_ms(QEMU_CLOCK_VIRTUAL) + s->regs.busy_delay);
+        }
+        break;
+    case BUSY_DELAY:
+        s->regs.busy_delay = val;
+        break;
+    default:
+        break;
+    }
+}
+
+static const MemoryRegionOps pcie_testdev_io_ops = {
+    .read = pcie_testdev_read,
+    .write = pcie_testdev_write,
+    .endianness = DEVICE_LITTLE_ENDIAN,
+};
+
+static void pcie_testdev_realize(PCIDevice *pcie_dev, Error **errp)
+{
+    PCIeTestDevState *s = PCI_EXPRESS_TEST_DEV(pcie_dev);
+
+    memory_region_init_io(&s->membar, OBJECT(s), &pcie_testdev_io_ops, s,"pci-testdev-membar", 65536);
```

Comment:
Please try to use consistent spacing. An auto-formatter (like clang-format) might be helpful. IIRC I've used this config file with some success in the past: https://gist.github.com/elmarco/aa5e0b23567f46fb7f0e73cde586a0c1

Reply:
Done, thanks!